libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java

```diff
@@ -615,14 +606,31 @@ private static boolean isTriviallyAllowed(Class<?> requestingClass) {
             logger.debug("Entitlement trivially allowed: no caller frames outside the entitlement library");
             return true;
         }
-        if (systemModules.contains(requestingClass.getModule())) {
+        if (isSystemModule(requestingClass.getModule())) {
             logger.debug("Entitlement trivially allowed from system module [{}]", requestingClass.getModule().getName());
             return true;
         }
         logger.trace("Entitlement not trivially allowed");
         return false;
     }
 
+    /**
+     * Determines if the given {@link Module} is a system module. That is, a module that is included as part of the Java runtime.
+     */
+    private static boolean isSystemModule(Module module) {
+        if (module.getLayer() == null) {
+            // This is an unnamed module, so not a system module
+            return false;
+        }
+
+        ModuleReference moduleReference = module.getLayer()
+            .configuration()
+            .findModule(module.getName())
+            .map(ResolvedModule::reference)
+            .orElse(null);
+        return module == PolicyManager.class.getModule() || (moduleReference != null && systemModules.contains(moduleReference));
```

Comment:
If it's the entitlements module we could return earlier? Then we don't need to do the reference unwrapping.